grpc/errors.go

```diff
@@ -63,59 +70,86 @@ func unwrapError(err error, md metadata.MD) error {
 	if !ok {
 		return err
 	}
+
+	inErrMsg := status.Convert(err).Message()
 	if len(errTypeStrs) != 1 {
 		return berrors.InternalServerError(
-			"multiple errorType metadata, wrapped error %q",
-			unwrappedErr,
+			"multiple 'errortype' metadata, wrapped error %q",
+			inErrMsg,
 		)
 	}
 
-	errType, decErr := strconv.Atoi(errTypeStrs[0])
+	inErrType, decErr := strconv.Atoi(errTypeStrs[0])
 	if decErr != nil {
 		return berrors.InternalServerError(
 			"failed to decode error type, decoding error %q, wrapped error %q",
 			decErr,
-			unwrappedErr,
+			inErrMsg,
 		)
 	}
-	outErr := berrors.New(berrors.ErrorType(errType), unwrappedErr)
-
-	subErrsJSON, ok := md["suberrors"]
-	if !ok {
-		return outErr
-	}
-	if len(subErrsJSON) != 1 {
-		return berrors.InternalServerError(
-			"multiple suberrors metadata, wrapped error %q",
-			unwrappedErr,
+	inErr := berrors.New(berrors.ErrorType(inErrType), inErrMsg)
+	var inBoulderErr *berrors.BoulderError
+	if !errors.As(inErr, &inBoulderErr) {
+		return fmt.Errorf(
+			"expected type of unwrappedErr to be %T got %T: %q",
+			inBoulderErr,
+			inErr,
+			inErr.Error(),
 		)
 	}
 
-	var suberrs []berrors.SubBoulderError
-	err2 := json.Unmarshal([]byte(subErrsJSON[0]), &suberrs)
-	if err2 != nil {
-		return berrors.InternalServerError(
-			"error unmarshaling suberrs JSON %q, wrapped error %q",
-			subErrsJSON[0],
-			unwrappedErr,
-		)
+	var decodedErrWithSubErrs *berrors.BoulderError
+	subErrorsVal, ok := md["suberrors"]
+	if ok {
+		if len(subErrorsVal) != 1 {
+			return berrors.InternalServerError(
+				"multiple 'suberrors' in metadata, wrapped error %q",
+				inErrMsg,
+			)
+		}
+
+		var subErrs []berrors.SubBoulderError
+		unmarshalErr := json.Unmarshal([]byte(subErrorsVal[0]), &subErrs)
+		if unmarshalErr != nil {
+			return berrors.InternalServerError(
+				"JSON unmarshaling 'suberrors' %q, wrapped error %q: %s",
+				subErrorsVal[0],
+				inErrMsg,
+				unmarshalErr,
+			)
+		}
+		decodedErrWithSubErrs = inBoulderErr.WithSubErrors(subErrs)
 	}
 
-	var berr *berrors.BoulderError
-	if errors.As(outErr, &berr) {
-		outErr = berr.WithSubErrors(suberrs)
-	} else {
-		return fmt.Errorf(
-			"expected type of outErr to be %T got %T: %q",
-			berr, outErr,
-			outErr.Error(),
-		)
+	var decodedRetryAfter int64
+	retryAfterVal, ok := md["retryafter"]
+	if ok {
+		if len(retryAfterVal) != 1 {
+			return berrors.InternalServerError(
+				"multiple 'retryafter' in metadata, wrapped error %q",
+				inErrMsg,
+			)
+		}
+		var parseErr error
+		decodedRetryAfter, parseErr = strconv.ParseInt(retryAfterVal[0], 10, 64)
+		if parseErr != nil {
+			return berrors.InternalServerError(
+				"parsing 'retryafter' as int64, wrapped error %q, parsing error: %s",
+				inErrMsg,
+				parseErr,
+			)
+		}
 	}
-	return outErr
+
+	if decodedErrWithSubErrs != nil {
+		decodedErrWithSubErrs.RetryAfter = decodedRetryAfter
+		return decodedErrWithSubErrs
+	}
+	inBoulderErr.RetryAfter = decodedRetryAfter
```

Comment:
There's a slight inconsistency between how we handle RetryAfter and how we handle SubErrors: for one we assign directly to the field; for the other we call `.WithSubErrors()` and create a new object.

The API design of BoulderError is a little funny, since it has this accessor to add SubErrors; but it also allows unmediated access to all its fields since they are all exported. I can't see discussion of this in https://github.com/letsencrypt/boulder/pull/4227. Maybe it was a convenience method?

The fields have to stay exported, because they are read and set from a different package (grpc). Given we're okay mutating `inBoulderErr` to set the RetryAfter, I think it would be reasonable to also mutate when adding SubErrors (rather than call WithSubErrors). That would allow getting rid of the `decodedErrWithSubErrs` intermediate value and simplifying some of the code paths here.

Reply:
Neat! This is much cleaner, thanks!